### DISCUSSION TITLE
Forslag til standard struktur for attributter

### DISCUSSION THREAD

**Discussion by eirikbroen** [NorskHelsenett/Tillitsrammeverk#22] [General]
I utprøvning har vi har god nytte av å spesifisere attributter på lik måte på følgende form

```javascript
{
  attributeName: {
    id: Verdi fra kodeverk [obligatorisk]
    name: Tekstlig beskrivelse av verdi [obligatorisk]
    system: OID for kodeverk [obligatorisk]
    assigner: Referanse til ansvarlig for kodeverk 
  }
}
```

Eksempel

```javascript
{
  "point_of_care_id": {
    "id": 984630492
    "name": "OSLO KOMMUNE HELSEETATEN LEGEVAKTEN STORGATA"
    "system": "urn:oid:2.16.578.1.12.4.1.101"
    "assigner": "https://www.brreg.no/"  
  }
}
```

Veldig praktisk å kunne forholde seg til en fast struktur for attributtbeskrivelser både for avgivende og mottagende part. I eksempelet over kan vi kanskje også unngå å sende både _point_of_care_id_ og _point_of_care_name_ hvis _name_ er et obligatorisk felt

**Comment by steinarnoem** [comment]
Jeg liker forslaget ditt godt! Spesielt synes jeg "assigner" elementet er interessant.

I dag har vi egentlig strukturert informasjonselementene litt forskjellig, f.eks. skiller vi mellom informasjon om virksomheter og kodeverk.

For virksomheter med org.nr. er det foreslått en slik struktur:
````JSON
"legal_entity": {
	"id": "921592761",
	"name": "Lege Leif Lagesen ENK",
	"oid": "x.x.x.x.x.x"
}
````

Mens, for kodeverk er det foreslått en slik struktur:
````JSON
"facility_type":{
	"text": "Sykepleietjeneste",
	"code": "KP02",
	"oid": "x.x.x.x.x.8663" //volven
}
````

Jeg synes strukturen som du foreslår er knakende god, men har et par meninger om navngiving:
Jeg tror vi har snakket om tre forskjellige typer informasjonselementer:
1. Identiteter
2. Verdier basert på kodeverk
3. Fritekst
Kanskje bør vi ha tre forskjellige strukturer for å uttrykke de forskjellige typer informasjonselementer?

Jeg synes at verdien for en kode (f.eks. "KP02" for "Sykepleietjeneste" burde ha claim-name: "code", mens identifikatorer, som f.eks et organisasjonsnummer eller et F-nummer burde ha claim-name: "id", ettersom verdien identifiserer en person (fysisk eller juridisk), mens en kode peker på et konsept/noe mer generisk.

I tillegg synes jeg at elementet som angir tekstlig verdi bør hete "name" hvis det faktisk er et navn, mens det bør hete "text" hvis det er en verdi som peker på et konsept/klassifisering (fra et kodeverk). 

Hva med noe sånt:
````JSON
"person": {
	"id": "xxxxxxxx",
	"name": "navn på juridisk eller fysisk person",
	"system": "x.x.x.x.x.x", //oid for F-nr/D-nr/Org.nr
        "authority": "www.brreg.no" //eller www.skatteetaten.no
}

"verdi_fra_kodeverk":{
	"code": "en_kode",
	"text": "en_beskrivende_tekst",
	"system": "oid kode for system",
	"assigner" "den_ansvarlige_for_kodeverket"
}

"fritekst": "verdi",

````

**Reply by OK-SverreM** [reply]
Bra forslag, Steinar!

**Reply by eirikbroen** [reply]
Liker godt forslaget ditt @steinarnoem! Enig i at det kan være fornuftig å ha to ulike typer for identiteter og verdi fra kodeverk. Når det gjelder felt for fritekst så er jeg mer usikker. På den ene siden er det lurt å ha med en type struktur som kan benyttes for utvidelser osv, men på den andre siden litt redd for at vi legger opp til å kunne sende verdier det blir vanskelig å validere. Kan det være en tanke å ha med _type_ på fritekst? Noe slikt:

```javscript
"fritekst": {
    "type": "NUMERIC|ALPHANUMERIC|DECIMAL",
    "value": "en verdi" 
}
```

**Reply by richardhus** [reply]
Hei
Forslaget fra @eirikbroen med tillegg fra @steinarnoem som støttes av @OK-SverreM er veldig i tråd med det vi har lagt til grunn internt i Helse Sør-Øst og som er brukt i begrenset utprøving/prøvedrift frem til nå. Det er veldig forutsigbart for maskinell behandling at man har en felles og forutsigbar struktur som er så lik som mulig. Vi har derfor praktisert å bruke grunnstrukturen med id, name, system og assigner uavhengig av om det er identifikator eller kodeverk som er benyttet og så har vi heller lagt til code som et attributt når vi har et kodeverk med tilhørende koder som ikke kan sies å utgjøre en reell identifikator for kodeverket. De fleste kodeverkene har gjerne helt unike id-er i databaser de er lagret i som ikke eksponeres for sluttbrukere eller inngår i kodeverket, men det er jo da typisk ikke identifikatorer som er like på tvers av virksomheter/applikasjoner når de har en lokal kopi av kodeverket, så den er ikke aktuell å formidle utenfor applikasjonen/virksomheten.

Hva som er kodeverk og hva som er identifikatorer kan også være vanskelig å vite i noen tilfeller, et eksempel kan være register over enheter i spesialisthelsetjenesten (RESH). Der har enhetene en unik ID (RESHID), så har de gjerne et (valgfritt) kortnavn, som brukes likt som koder i kodeverk og ofte omtales av sluttbrukere som kortkode eller rekvirentkode og et tilhørende navn. Vi klarer oss antakelig fint uten å formidle kortnavnet fra RESH i denne sammenhengen, så her tenker vi at dette er en ren identifikator. Jeg tok dette med her bare for å illustrere at det finnes noen eksempler man kanskje må ta høyde for der både ID og kode kan være aktuelle samtidig.

Et annet og mer relevant eksempel er grunnlaget for tilgang som vi ønsker å formidle som DIPS-konsument til alle kilder når dette er nødvendig for å få tilgang til kildene. Da har man en identifikator på den lokale beslutningen som identifiserer den spesifikke beslutningen av tilgang gjort av et gitt helsepersonell for å få tilgang til en spesifikk pasient. Sammen med denne identifikatoren ønsker vi å formidle koden fra et kodeverk over mulige årsaker til tilgangen som i dette tilfelle er den overordnede årsaken til denne spesifikke tilgangen. I DIPS omtales dette kodeverket som beslutningsmaler. Beslutningsmalen har en kode i tillegg til en intern databaseidentifikator (som er uinteressant utenfor DIPS-instansen og derfor ikke aktuell å ta med her heldigvis), et navn og en mer utdypende tekstlig begrunnelse. Alle disse er egentlig relevante for eksterne virksomheter som selv ikke kjenner til dette kodeverket og for innbyggere som heller ikke kjenner det. Derfor har vi også lagt til ytterligere attributter enn code akkurat ved formidling av det spesifikke grunnlaget for tilgangen fra lokalt EPJ i vår struktur internt i Helse Sør-Øst.

Men for beslutninger trenger vi altså både id på den spesifikke beslutningen og koden fra kodeverket for beslutningsmaler. Så tenker jeg at dette er et veldig spesialtilfelle og derfor ikke kan inngå i grunnstrukturen, men vi har så langt i internt sikkerhetsbevis for Helse Sør-Øst klart oss veldig godt med å formidle både kodeverk og identifikatorer på lik måte via bare id, name, system og assigner, og så heller legge til ekstra attributter de få stedene det er behov for noe mer.

Gir det egentlig reell og nyttig merverdi å ha ulik navngiving/struktur på identifikatorer og kodeverk er mitt spørsmål?

Mer om behovet for spesiell struktur ved formidling av grunnlaget for tilgang som gjør at dette uansett må særbehandles fra en grunnstruktur:

Den spesifikke beslutningen av tilgang i lokalt EPJ består i tillegg til det som er beskrevet over av flere faktorer enn beslutningsmalen som medførte at man endte opp med å få tilgang. For eksempel tilhørighet til organisasjonsenhet/avdeling for pasienten og tidspunktet for tilgangen. Tidspunktet er viktig fordi det i mange tilfeller bekrefter at tilgangen samsvarer med akseptert tidsperiode for tilgang ut fra pasientadministrative registreringer, og det gjør det enklere for innbyggere å forstå hvilken opplevd kontakt med helsevesenet som tilgangen har opphav fra. DIPS konkatenerer disse faktorene sammen i et tekstbasert felt som lagres og logges tilhørende den spesifikke beslutningen om tilgang. Jeg mener det har høy verdi for både innbygger og andre konsumenter av logginformasjon hos kilder å motta all denne informasjonen, selv om mye av den er konkatenerte verdier i en tekststreng pr. nå. Strukturen vi har lagt opp til for å formidle grunnlaget for tilgang nå i Helse Sør-Øst ser omtrent slik ut:
hso:purpose-local:id - identifikator på spesifikk beslutning om tilgang mellom helsepersonell og pasient - vil være ulik pr. beslutning
hso:purpose-local:name - navn på beslutningmal
hso:purpose-local:system - uri for kodeverk for beslutningsmal
hso:purpose-local:assigner - forvalter/utsteder av koder i kodeverk for beslutningsmaler (DIPS)
hso:purpose-local:description - beskrivelse (konkatenert tekst fra DIPS) av beslutningen
hso:purpose-local:reason - fritekst fra sluttbruker som forklarer tilgangen spesifikt (obligatorisk ved egenbesluttet tilgang)
hso:purpose-local:userSelected - Indikerer om det er en egenbesluttet tilgang eller ikke (true/false)
hso:purpose-local:code - kode på beslutningsmal som vil være lik mellom alle som har samme kodeverk og der samme beslutningsmal er brukt

Det er helt klart grunn til å diskutere om system/assigner alltid må være koblet til id og ikke til code-verdien slik vi har lagt til grunn her. Eventuelt om forslaget til @steinarnoem over løser dette.

I tillegg bidrar også følgende attributter til å dokumentere grunnlaget for tilgangen:
hso:healthcareservice:* - helsehjelpstjeneste/fagområde som ytes til pasienten i forbindelse med tilgangen
hso:subject:child-organization:* - arbeidsstedet (geografisk tilhørighet) til helsepersonell
hso:subject:functional-role:* - funksjonelt orientert rollebegrep som indikerer hvilket ansvar og hvilke oppgaver helsepersonellet har
hso:subject:facility:* - mest detaljert organisasjonstilhørighet som oppgis for helsepersonell (typisk avdeling/seksjon)
hso:resource:child-organization:* - behandlingsstedet pasientens behandles ved (geografisk tilhørighet)
hso:resource:facility:* - mest detaljert organisasjonstilhørighet for pasienten (typisk poliklinikk/sengepost)

stjerne (*) indikerer at vi her har id, name, system og assigner og ikke verken mer eller mindre.

**Comment by steinarnoem** [comment]
Feature-teamet har konkludert med at følgende struktur skal benyttes i informasjons- og datamodellen. 
Eventuelle endringer eller forfininger tas som issues, og denne diskusjonstråden låses.
Vi har opprettet [Issue #38](https://github.com/NorskHelsenett/Tillitsrammeverk/issues/38) for å ta strukturen inn i spesifikasjonen.

````JSON
"person (juridisk/fysisk)": {
	"id": "xxxxxxxx",
	"name": "navn på juridisk eller fysisk person",
	"system": "x.x.x.x.x.x", //oid for F-nr/D-nr/Org.nr
    "authority": "www.brreg.no" //eller www.skatteetaten.no
}

"verdi_fra_kodeverk":{    
	"code": "en_kode",
	"text": "en_beskrivende_tekst",
	"system": "oid kode for system", 
	"assigner": "den_ansvarlige_for_kodeverket"
}

"fritekst": {
    "type": "NUMERIC|ALPHANUMERIC|DECIMAL",
    "value": "en verdi" 
}

"hendelse_referanse":{
    "id": "xxxxxxxx",
	"verdi_fra_kodeverk": {
        "code": "XX",
        "text": "en_beskrivende_tekst",
        "system": "x.x.x.x.x.x",
        "assigner": "den_ansvarlige_for_kodeverket"
    },
	"person (juridisk/fysisk)": {
        "id": "xxxxxxxx",
        "name": "navn på juridisk eller fysisk person",
        "system": "x.x.x.x.x.x", //oid for F-nr/D-nr/Org.nr
        "authority": "www.brreg.no" //eller www.skatteetaten.no
    },
    "fritekst":{
        "type": "NUMERIC|ALPHANUMERIC|DECIMAL",
        "value": "en verdi" 
    }
}

Eks 1: Tilgangsbeslutning/hjemmelsgrunnlag
"hendelse_referanse":{
    "id": "xxxxxxxx",
	"verdi_fra_kodeverk": {
        "code": "XX",
        "text": "en_beskrivende_tekst",
        "system": "x.x.x.x.x.x",
        "assigner": "den_ansvarlige_for_kodeverket"
    },
   "fritekst":{
        "type": "NUMERIC|ALPHANUMERIC|DECIMAL",
        "value": "en verdi" 
    }
}

Eks 2: Tilgangsbeslutning/hjemmelsgrunnlag (ikke unntak)
"hendelse_referanse":{
    "id": "xxxxxxxx",
	"person (juridisk/fysisk)": {
        "id": "xxxxxxxx",
        "name": "navn på juridisk eller fysisk person",
        "system": "x.x.x.x.x.x", //oid for F-nr/D-nr/Org.nr
        "authority": "www.brreg.no" //eller www.skatteetaten.no
    },
    "fritekst":{
        "type": "NUMERIC|ALPHANUMERIC|DECIMAL",
        "value": "en verdi" 
    }
}

````
